### PR TITLE
Fix to non-deterministic behaviour by removing long-range side effects

### PR DESCRIPTION
**Pull request by williamjameshandley**
# Description

This PR achieve two things
1) It streamlines the CI so that we test [osx,windows]x[pip,conda]x[3.10] + [linux]x[pip,conda]x[3.8,3.9,3.10]. This reduces build times since osx and windows are reduced by a factor of 3, and expands the (much faster) linux builds to include 3.9.
2) We aim to remove all side effects from pandas plotting functions

(2) is an insidious bug which has been affecting all current PRs non-deterministically. Tests pass on local machines, but when lifted to the CI servers sometimes break in certain functions. The hypothesis is that pandas plotting is very inconsistent, for example `.plot.box()` creates a new axis, whilst `.boxplot()` uses the 'current axis' with e.g. `plt.gca()`. It is the latter functions which cause the problem by potentially getting a current axis somewhere else in the test suite, which depends on how fast/parallel/ordered the builds are. Failing tests can thus arise in either function depending on the current axis and the effect of both things writing to it simultaneously, so apparently healthy functions can fail. We aimed to control this by copious `plt.close()` commands, but this only partially solves the issue since presumably a new axis can be created in between the last `plt.close()` and the next `plt.gca()`.

The solution is to go through the test suite and find any functions which have `plt.gca()` functionality, and to explicitly pass them a newly created axis. I have done this, and removed all `plt.close()` commands.

This has been re-visited in #235 #221 #222

# Checklist:

- [X] I have performed a self-review of my own code
- [X] My code is PEP8 compliant (`flake8 anesthetic tests`)
- [X] My code contains compliant docstrings (`pydocstyle --convention=numpy anesthetic`)
- [X] New and existing unit tests pass locally with my changes (`python -m pytest`)
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: tests/test_plot.py [hunk 1]
```diff
@@ -21,6 +21,13 @@
 from scipy.interpolate import interp1d
 
 
+@pytest.fixture(autouse=True)
+def close_figures_on_teardown():
+    plt.figure()
```
[the remark this answers is not in the document]
Comment (lukashergt) [reply tests/test_plot.py]: Huh, a rerun of the tests after merging master failed... The joy of non-deterministic bugs...

**Comment by lukashergt** [comment]
Tests failed again with the now well-known but not understood limits mismatch in AreaPlot. Maybe #244 can resolve this...

**Review by lukashergt — APPROVED** [review]
Looks like everything passes fine, so hopefully this did the trick. Thanks @Ormorod and @williamjameshandley for the pandas deep dive!

Happy to approve this as is, but maybe one point we want to add:

Would we prefer creating new figure and axes at the start of `test_AreaPlot` (which is where things kept going awry, if not where the problems originated...)?

**Comment by Ormorod** [comment]
> ...
> Would we prefer creating new figure and axes at the start of `test_AreaPlot` (which is where things kept going awry, if not where the problems originated...)?

agree

File: tests/test_weighted_pandas.py [hunk 2]
```diff
@@ -809,42 +805,37 @@ def test_HexBinPlot(mcmc_df, mcmc_wdf):
 
 
 def test_AreaPlot(mcmc_df, mcmc_wdf):
-    axes_wdf = mcmc_wdf.plot.area()
-    axes_df = mcmc_df.plot.area()
+    fig, (ax1, ax2) = plt.subplots(1, 2)
+    axes_df = mcmc_df.plot.area(ax=ax1)
+    axes_wdf = mcmc_wdf.plot.area(ax=ax2)
```
Comment (lukashergt) [inline tests/test_weighted_pandas.py:810]: I've created new figure and axes in `test_AreaPlot`.

**Review by lukashergt — APPROVED** [review]

**Comment by Ormorod** [comment]
The `AreaPlot` issue seems to have reappeared...

**Comment by williamjameshandley** [comment]
Yeah -- I spotted this in https://github.com/williamjameshandley/anesthetic/pull/243#issuecomment-1407439986. There are still a couple of uses of `gcf` and `gca` in the `hist` and `bootstrap` plots, but even if this were the issue, it wouldn't be consistent with our previous hypothesis

**Pull request merged**